src/networks/evm/client.ts

```
@@ -40,164 +43,234 @@ export class EthereumClient extends Client {
-
-    return name;
-  }
-
-  public async transfer(to: string, amount: string): Promise<string> {
-    const from = await this.getAddress();
-    const nonce = await this.web3.eth.getTransactionCount(from);
-    const gas = await this.web3.eth.estimateGas({ from, to, value: amount });
-    const gasPrice = await this.web3.eth.getGasPrice();
-    const signed = await this.web3.eth.signTransaction({
-      from,
-      to,
-      value: amount,
-      nonce,
-      gas,
-      gasPrice: BigInt(Math.ceil(Number(gasPrice) * this.gasMultiplier)).toString(),
-    });
-
-    const receipt = await this.web3.eth.sendSignedTransaction(signed.raw);
-    const block = await this.web3.eth.getBlock(receipt.blockNumber);
-
-    let timestamp;
-    if (typeof block.timestamp == 'string') {
-      timestamp = parseInt(block.timestamp);
-    } else {
-      timestamp = block.timestamp;
-    }
-
-    let status = TxStatus.Completed;
-    if (!receipt.status) {
-      status = TxStatus.Error;
-    }
-
-    const nativeTx = await this.web3.eth.getTransaction(receipt.transactionHash);
-    convertTransaction(nativeTx, timestamp, status);
-
-    return receipt.transactionHash;
+    return this.contractCallRetry(this.token, tokenAddress, 'name');
   }
 
   public async decimals(tokenAddress: string): Promise<number> {
     let res = this.tokenDecimals.get(tokenAddress);
     if (!res) {
       try {
         this.token.options.address = tokenAddress;
-        res = Number(await this.token.methods.decimals().call());
+        res = Number(await await this.contractCallRetry(this.token, tokenAddress, 'decimals'));
         this.tokenDecimals.set(tokenAddress, res);
       } catch (err) {
-        console.log(`Cannot fetch decimals for the token ${tokenAddress}, using default (18). Reason: ${err.message}`);
+        console.warn(`[SupportJS] Cannot fetch decimals for the token ${tokenAddress}, using default (18). Reason: ${err.message}`);
         res = 18;
```

I am not sure that we should use default value for decimals. It looks quite dangerous in my opinion.

---

Agree with you. Let's throw an error instead. Fixed in https://github.com/zkBob/zeropool-support-js/pull/9/commits/96c6edc93af0da09efec3308f6df99d4df414046